isarn-sketches-java/src/main/java/org/isarnproject/sketches/java/TDigest.java

```diff
@@ -0,0 +1,761 @@
+     * the larger as the result.
+     * Note this means either (ltd) or (rtd) will be modified.
+     * @param ltd a t-digest
+     * @param rtd another t-digest
+     * @return if ltd has larger mass, then returns <pre>ltd.merge(rtd)</pre>,
+     * otherwise <pre>rtd.merge(ltd)</pre>
+     */
+    public static TDigest merge(TDigest ltd, TDigest rtd) {
+        if (ltd.size() < rtd.size()) return merge(rtd, ltd);
+        if (rtd.size() == 0) return ltd;
+        if (rtd.size() == 1) {
+            ltd.update(rtd.cent[0], rtd.mass[0]);
+            return ltd;
+        }
+        if (rtd.mass() < ltd.mass()) {
+            ltd.merge(rtd);
+            return ltd;
+        } else {
+            rtd.merge(ltd);
+            return rtd;
+        }
+    }
+
+    /**
+     * Sketch data using a t-digest with default compression and maximum discrete tracking.
+     * @param data the data to sketch
+     * @return a t-digest sketch of the data
+     */
+    public static TDigest sketch(double[] data) {
+        return sketch(data, COMPRESSION_DEFAULT, 0, INIT_SIZE_DEFAULT);
+    }
+
+    /**
+     * Sketch data using a t-digest.
+     * maxDiscrete defaults to zero.
+     * @param data the data to sketch
+     * @param compression sketching compression setting. Higher = more compression.
+     * Must be &gt; 0.
+     * @return a t-digest sketch of the data
+     */
+    public static TDigest sketch(double[] data, double compression) {
+        return sketch(data, compression, 0, INIT_SIZE_DEFAULT);
+    }
+
+    /**
+     * Sketch data using a t-digest.
+     * @param data the data to sketch
+     * @param compression sketching compression setting. Higher = more compression.
+     * Must be &gt; 0.
+     * @param maxDiscrete maximum number of unique discrete values to track. Must be &ge; 0.
+     * If this number of values is exceeded, the sketch will begin to operate in 
+     * normal continuous mode.
+     * @return a t-digest sketch of the data
+     */
+    public static TDigest sketch(double[] data, double compression, int maxDiscrete) {
+        return sketch(data, compression, maxDiscrete, INIT_SIZE_DEFAULT);
+    }
+
+    /**
+     * Sketch data using a t-digest.
+     * @param data the data to sketch
+     * @param compression sketching compression setting. Higher = more compression.
+     * Must be &gt; 0.
+     * @param maxDiscrete maximum number of unique discrete values to track. Must be &ge; 0.
+     * If this number of values is exceeded, the sketch will begin to operate in 
+     * normal continuous mode.
+     * @param sz initial capacity to use for internal arrays. Must be &gt; 0.
+     * @return a t-digest sketch of the data
+     */
+    public static TDigest sketch(double[] data, double compression, int maxDiscrete, int sz) {
+        TDigest td = empty(compression, maxDiscrete, sz);
+        for (double x: data) td.update(x, 1.0);
+        if (td.size() > maxDiscrete) td.recluster();
```

Comment:
isn't this redundant?

Reply:
The reclustering is taken from the original paper, as a heuristic for making sure the final output is compact. It is not necessarily redundant in the sense that re-clustering might be warranted but may not have "quite happened yet" as part of the normal update logic.

Reply:
OK, so if I'm reading `update(x,w)` correctly, if your final update leaves `nclusters` both over `maxDiscrete` and over `R()` this will be redundant, but not at other times -- and reclustering is cheap so there's no downside to this belt-and-suspenders approach.  Thanks for the clarification (I had missed a couple of points while scrolling around this diff).